url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/core/RWLock.java

```diff
@@ -0,0 +1,143 @@
+// Spatial Index Library
+//
+// Copyright (C) 2002  Navel Ltd.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License aint with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+//
+// Contact information:
+//  Mailing address:
+//    Marios Hadjieleftheriou
+//    University of California, Riverside
+//    Department of Computer Science
+//    Surge Building, Room 310
+//    Riverside, CA 92521
+//
+//  Email:
+//    marioh@cs.ucr.edu
+
+// Readers/Writers lock by Allen Holub
+
+package spatialindex.core;
+
+import java.util.LinkedList;
+
+public class RWLock {
+	private int active_readers;
+	private int waiting_readers;
+	private int active_writers;
+
+	private final LinkedList<Object> writer_locks = new LinkedList<Object>();
+
+	/**
+	 * if no one is writing and !write_lock, then allow to read
+	 */
+	public synchronized void read_lock() {
+		if (active_writers == 0 && writer_locks.size() == 0)
+			++active_readers;
+		else {
+			++waiting_readers;
+			try {
+				wait();
+			} catch (InterruptedException e) {
+			}
+		}
+	}
+
+	/**
+	 * 
+	 * @return whether the read_lock is unblock
+	 */
+	public synchronized boolean read_lock_noblock() {
+		if (active_writers == 0 && writer_locks.size() == 0) {
+			++active_readers;
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * if no one reader is reading, then unblock the read_block and notify
+	 * writers
+	 */
+	public synchronized void read_unlock() {
+		if (--active_readers == 0)
+			notify_writers();
+	}
+
+	/**
+	 * if it is safe for write, then return. else wait until it is safe to write
+	 */
+	public void write_lock() {
+		Object lock = new Object();
+		synchronized (lock) {
+			synchronized (this) {
+				boolean okay_to_write = writer_locks.size() == 0
+						&& active_readers == 0 && active_writers == 0;
+				if (okay_to_write) {
+					++active_writers;
+					return; // the "return" jumps over the "wait" call
+				}
+
+				writer_locks.addLast(lock);
+			}
+			try {
+				lock.wait();
+			} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)